### PR TITLE
29 create initial strike log postgres table and unit test

### LINKED ISSUE #29: Create Initial StrikeLog Postgres Table and Unit Test
### Subject of the issue

Create an initial strike log database table to store strikes.  And initial database indexes, etc... to facilitate searching, sorting etc...  Also create server side unit test.

### Steps to reproduce
-
-
-

### Image

Add any images that might be helpful to understand the problem.

### PR DESCRIPTION
**Pull request by smelserm**
Fixes #29 

### Changes proposed in this pull request
- Add database Schema StrikeLog 
- Add initial Strike Table
- Add rest tier for basic CRUD on Strike Table
- Add unit test to test CRUD and Row Level Security, etc...
- Add unit server test back to CircleCI

### Notes on how to test this pull request
- CircleCI actually does a "make test" which is the best way to test this PR
- For ambitious PR reviewers: Pull the PR, do a "make test" in root of strikelog

### Checklist
- [x] Code compiles correctly
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing

### CODE REVIEW

File: setup/0003_Add_Strike_Table.up.sql [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    striketime timestamp without time zone NOT NULL,
+    PRIMARY KEY (rowid),
+    UNIQUE (id,rowid,updateid)
+);
+
+CREATE INDEX strike_lastmodified_idx ON strike (writetime);
+CREATE INDEX strike_entityname_idx ON strike (entityname);
+CREATE INDEX strike_striketime_idx ON strike (striketime);
+
+--Begin common RLS POLICY for strike
```
Comment (bofabcd) [inline setup/0003_Add_Strike_Table.up.sql]: Remove policy from here, it is added by set_permissions.sql

File: setup/0002_Create_StrikeLog_Schema.up.sql [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+SET standard_conforming_strings = on;
+SET check_function_bodies = false;
+SET client_min_messages = warning;
+SET row_security = off;
+
+CREATE SCHEMA IF NOT EXISTS strikelog;
+ALTER SCHEMA strikelog OWNER TO postgres;
+
+GRANT USAGE ON SCHEMA strikelog TO strikelog;
+GRANT USAGE ON SCHEMA strikelog TO strikelogfuncs;
```
Comment (bofabcd) [inline setup/0002_Create_StrikeLog_Schema.up.sql]: no need for strikelogfuncs

File: setup/0002_Create_StrikeLog_Schema.up.sql [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+       , CASE WHEN _sort <> '' THEN 'ORDER BY ' || _sort END
+       , CASE WHEN filters::jsonb->>'limit' IS NOT NULL THEN 'LIMIT $5' END
+       , CASE WHEN filters::jsonb->>'offset' IS NOT NULL THEN 'OFFSET $6' END
+    );
+
+	RETURN _query;
+END        
+$$;
+
+ALTER FUNCTION strikelog.get_query_with_params(sqlSelect text, sqlWhere text, filters jsonb)
```
Comment (bofabcd) [inline setup/0002_Create_StrikeLog_Schema.up.sql]: Lines 81 - 84 should be removed, this is handled bt set_permissions.sql

File: setup/0003_Add_Strike_Table.up.sql [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        (SELECT ARRAY(SELECT jsonb_array_elements_text(($1::jsonb->>'readgroups')::jsonb))),
+        (SELECT ARRAY(SELECT jsonb_array_elements_text(COALESCE(($1::jsonb->>'writegroups')::jsonb, ($1::jsonb->>'readgroups')::jsonb)))),
+        ($1::jsonb->>'classification')::jsonb,
+        ($1::jsonb->>'entityname')::text, ($1::jsonb->>'striketime')::timestamp
+    )
+    RETURNING to_jsonb(strike.*);
+END
+$$;
+
+ALTER FUNCTION strikelog.insert_or_update_strike(params jsonb)
```
Comment (bofabcd) [inline setup/0003_Add_Strike_Table.up.sql]: Lines 76 - 77 should be removed, this is handled by set_permissions.sql

File: setup/0003_Add_Strike_Table.up.sql [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    _sql := strikelog.get_query_with_params(_sqlSelect, _where, $1);
+
+    RETURN QUERY EXECUTE _sql
+    USING ($1::jsonb->>'id')::uuid, ($1::jsonb->>'userid')::text, ($1::jsonb->>'modifiedafter')::timestamp,
+    ($1::jsonb->>'modifiedbefore')::timestamp, ($1::jsonb->>'limit')::int,
+    ($1::jsonb->>'offset')::int, ($1::jsonb->>'entityname')::text;
+END        
+$$;
+
+ALTER FUNCTION strikelog.get_strikes(filters jsonb)
```
Comment (bofabcd) [inline setup/0003_Add_Strike_Table.up.sql]: Lines 105 - 106 should be removed, this is handled by set_permissions.sql

File: setup/0003_Add_Strike_Table.up.sql [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+$$;
+
+ALTER FUNCTION strikelog.set_deleted_strike(dataid uuid, deleted boolean)
+    OWNER TO strikelogfuncs;
+
+GRANT ALL ON FUNCTION strikelog.insert_or_update_strike(params jsonb) TO strikelog;
+GRANT ALL ON FUNCTION strikelog.get_strikes(filters jsonb) TO strikelog;
+GRANT ALL ON FUNCTION strikelog.set_deleted_strike(id uuid, deleted boolean) TO strikelog;
+
+GRANT SELECT,INSERT,DELETE,UPDATE ON TABLE strike TO strikelogfuncs;
```
Comment (bofabcd) [inline setup/0003_Add_Strike_Table.up.sql]: Lines 122 - 129 should be removed, this is handled by set_permissions.sql

File: setup/0003_Add_Strike_Table.up.sql (hunk 1 above)
Comment (smelserm) [reply setup/0003_Add_Strike_Table.up.sql]: I copied all this from CEM to IPT and then to here so it probably needs to be updated everywhere.  Good Catch though.

**Comment by smelserm** [comment]
Brian I think I cleaned up everything you ask for.  That sure makes things a lot cleaner!!

**Pull request merged**